sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodShouldBeNamedAccordingToSynchronicity.cs

```diff
@@ -72,24 +73,37 @@ public sealed class MethodShouldBeNamedAccordingToSynchronicity : SonarDiagnosti
                         return;
                     }
 
-                    var isTaskReturnType = (methodSymbol.ReturnType as INamedTypeSymbol)?.ConstructedFrom.DerivesFromAny(TaskTypes) ?? false;
-                    var hasAsyncSuffix = methodDeclaration.Identifier.ValueText.SplitCamelCaseToWords().LastOrDefault() == "ASYNC";
+                    var hasAsyncReturnType = HasAsyncReturnType(methodSymbol);
+                    var hasAsyncSuffix = HasAsyncSuffix(methodDeclaration);
 
-                    if (hasAsyncSuffix && !isTaskReturnType)
+                    if (hasAsyncSuffix && !hasAsyncReturnType)
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, methodDeclaration.Identifier.GetLocation(), RemoveAsyncSuffixMessage));
                     }
-                    else if (!hasAsyncSuffix && isTaskReturnType)
+                    else if (!hasAsyncSuffix && hasAsyncReturnType)
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, methodDeclaration.Identifier.GetLocation(), AddAsyncSuffixMessage));
                     }
-                    else
-                    {
-                        // do nothing
-                    }
                 },
                 SyntaxKind.MethodDeclaration);
 
+        private static bool HasAsyncReturnType(IMethodSymbol methodSymbol)
+        {
+            var returnSymbol = (methodSymbol.ReturnType as INamedTypeSymbol)?.ConstructedFrom;
+            if (returnSymbol == null)
+            {
+                return false;
+            }
+
+            return returnSymbol.DerivesFromAny(asyncReturnTypes) ||
+                   returnSymbol.IsAny(asyncReturnInterfaces) ||
```

Comment:
`.DerivesFromAny` checks also the equality between the type parameter and known type but, for interfaces, `.ImplementsAny` is not doing that and we have to check the equality using `IsAny`.

Reply:
hmm, this is confusing...

it would be nice to check where we're using `ImplementsAny` and find errors , but not sure we should open a ticket for it